File: 202.Happy_Number/HappyNumber.py
# Fast-Slow Pointer
# Pay more attention in Python, the // operator is used for integer division, which discards the fractional part.
# This is equivalent to / in Java when used with integers.
class Solution:
    def getNext(self, num):
        total = 0
        while num > 0:
            a = num % 10
            num = num // 10
            total += a * a
        return total

    def isHappy(self, n):
        slow = n
        fast = self.getNext(n)
        while fast != 1 and fast != slow:
            slow = self.getNext(slow)
            fast = self.getNext(self.getNext(fast))
        return fast == 1

File: 202.Happy_Number/HappyNumber.go
package main

// getNext and isHappy, these two function are just standalone functions.
func getNext(num int) int {
	total := 0
	for num > 0 {
		a := num % 10
		num = num / 10
		total += a * a
	}
	return total
}

func isHappy(n int) bool {
	slow := n
	fast := getNext(n)
	for fast != 1 && fast != slow {
		slow = getNext(slow)
		fast = getNext(getNext(fast))
	}
	return fast == 1
}